Apps/TimelineApp/feature/user/src/main/java/uk/ac/cam/cares/jps/user/SensorSettingFragment.java

```diff
@@ -0,0 +1,297 @@
+        });
+        initPermissions();
+
+        return binding.getRoot();
+    }
+
+
+    /**
+     * Called immediately after onCreateView has returned, but before any saved state has been restored in to the view.
+     *
+     * @param view The View returned by {@link #onCreateView(LayoutInflater, ViewGroup, Bundle)}.
+     * @param savedInstanceState If non-null, this fragment is being re-constructed
+     * from a previous saved state as given here.
+     */
+    @Override
+    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+
+        RecyclerView recyclerView = view.findViewById(R.id.sensors_recycler_view);
+        recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
+
+        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(),
+                LinearLayoutManager.VERTICAL);
+        recyclerView.addItemDecoration(dividerItemDecoration);
+
+        List<SensorItem> sensorItems = new ArrayList<>();
+        sensorItems.add(new SensorItem("Accelerometer", "Measures acceleration.", sensorManager.getSensorHandler(SensorType.ACCELEROMETER)));
+        sensorItems.add(new SensorItem("Gyroscope", "Tracks rotation rate.", sensorManager.getSensorHandler(SensorType.GYROSCOPE)));
+        sensorItems.add(new SensorItem("Magnetometer", "Detects magnetic fields.", sensorManager.getSensorHandler(SensorType.MAGNETOMETER)));
+        sensorItems.add(new SensorItem("Light", "Senses light levels.", sensorManager.getSensorHandler(SensorType.LIGHT)));
+        sensorItems.add(new SensorItem("Humidity", "Monitors air moisture.", sensorManager.getSensorHandler(SensorType.HUMIDITY)));
+        sensorItems.add(new SensorItem("Pressure", "Gauges atmospheric pressure.", sensorManager.getSensorHandler(SensorType.PRESSURE)));
+        sensorItems.add(new SensorItem("Gravity", "Detects gravity vector.", sensorManager.getSensorHandler(SensorType.GRAVITY)));
+        sensorItems.add(new SensorItem("Location", "Tracks GPS position.", sensorManager.getSensorHandler(SensorType.LOCATION)));
+        sensorItems.add(new SensorItem("Microphone", "Capture sound levels.", sensorManager.getSensorHandler(SensorType.SOUND)));
+
+        adapter = new SensorAdapter(sensorItems, this);
+        recyclerView.setAdapter(adapter);
+
+
+        Button toggleAllBtn = view.findViewById(R.id.toggle_all_btn);
+        toggleAllBtn.setOnClickListener(v -> {
+            // Toggle the state
+            allToggledOn = !allToggledOn;
+
+
+            for (SensorItem item : adapter.getSensorItems()) {
+                item.setToggled(allToggledOn); // Set the toggled state
+            }
+
+            // Notify the adapter to refresh the UI
+            adapter.notifyDataSetChanged();
+
+
+            toggleAllBtn.setText(allToggledOn ? "Toggle Off" : "Toggle All");
```

Comment:
it makes more sense to put `allToggledOn` in ViewModel and only update the `allToggledOn` here. Then listen to the `allToggledOn` live data to do all the ui update. Can refer to line153.